src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
+        } else /*if (state.isInstanceOf[Callbacks[T]]) */ {
+          val l = if (link ne null) link else new Link(target)
+          val p = l.promise(this)
+          if (p ne this) {
+            if (compareAndSet(state, l)) {
+              if (state ne Noop) p.dispatchOrAddCallbacks(p.get(), state.asInstanceOf[Callbacks[T]])
+            } else linkRootOf(p, l)
+          }
+        }
       }
-    }
   }
 
-  /** An already completed Future is given its result at creation.
-   *
-   *  Useful in Future-composition when a value to contribute is already available.
+  // Constant byte tags for unpacking transformation function inputs or outputs
+  final val Xform_noop          = 0
+  final val Xform_map           = 1
+  final val Xform_flatMap       = 2
+  final val Xform_transform     = 3
+  final val Xform_transformWith = 4
+  final val Xform_foreach       = 5
+  final val Xform_onComplete    = 6
+  final val Xform_recover       = 7
+  final val Xform_recoverWith   = 8
+  final val Xform_filter        = 9
+  final val Xform_collect       = 10
+  //final val Xform_failed        = 11
+
+    /* Marker trait
    */
-  object KeptPromise {
-    import scala.concurrent.Future
-    import scala.reflect.ClassTag
+  sealed trait Callbacks[-T] {
+    def submitWithValue(v: Try[T]): Unit
+  }
+
+  final class ManyCallbacks[-T](final val first: Callbacks[T], final val last: Callbacks[T]) extends Callbacks[T] {
+    // NOTE: This does grow the stack for *linked* callbacks which are transported across the links,
+    // these should normally be rather flat.
+    final def submitWithValue(v: Try[T]): Unit = {
+      first.submitWithValue(v)
+      last.submitWithValue(v)
+    }
+    override final def toString: String = "ManyCallbacks"
+  }
 
-    private[this] sealed trait Kept[T] extends Promise[T] {
-      def result: Try[T]
+  private[this] final val Noop = new Transformation[Nothing, Nothing](Xform_noop, null, InternalCallbackExecutor)
+
+  final class Transformation[-F, T] private[this] (
+    private[this] final var _fun: Any => Any,
+    private[this] final var _arg: AnyRef,
+    private[this] final val _xform: Byte
+  ) extends DefaultPromise[T]() with Callbacks[F] with Runnable with OnCompleteRunnable {
+    final def this(xform: Int, f: _ => _, ec: ExecutionContext) = this(f.asInstanceOf[Any => Any], ec.prepare(): AnyRef, xform.asInstanceOf[Byte])
+
+    // Gets invoked when a value is available, schedules it to be run():ed by the ExecutionContext
+    // submitWithValue *happens-before* run(), through ExecutionContext.execute.
+    // Invariant: _arg is `ExecutionContext`
+    // requireNonNull(v) will hold as guarded by `resolve`
+    override final def submitWithValue(resolved: Try[F]): Unit =
+      if (_xform != Xform_noop) {
+        val executor = _arg.asInstanceOf[ExecutionContext]
+        try {
+          _arg = resolved
+          executor.execute(this) // Safe publication of _arg = v (and _fun)
```

Comment:
Was `resolved` previously called `v`? Regardless, the comments need updating, as there is no `v` here

Reply:
@NthPortal Nice catch!